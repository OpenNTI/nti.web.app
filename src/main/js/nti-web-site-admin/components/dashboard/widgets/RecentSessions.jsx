import './RecentSessions.scss';
import React from 'react';
import PropTypes from 'prop-types';

import { scoped } from '@nti/lib-locale';
import { DateTime, Loading, Avatar } from '@nti/web-commons';
import { getService, User } from '@nti/web-client';
import { getLink } from '@nti/lib-interfaces';

const ANALYTICS = 'Analytics';
const SESSIONS_LINK = 'sessions';
const LIMIT = 10;
const PAGE_SIZE = 6;

const LABELS = {
	title: 'Recent Sessions',
	name: 'Name',
	value: '',
	noItems: 'No sessions found',
};

const t = scoped(
	'nti-web-site-admins.components.dashboard.widgets.recentsessions',
	LABELS
);

class Item extends React.Component {
	static propTypes = {
		item: PropTypes.object.isRequired,
	};

	renderImg() {
		const { item } = this.props;

		return <Avatar className="item-image" entity={item.user} />;
	}

	getDurationString(duration) {
		const hours = Math.floor(duration / (60 * 60));
		const minutes = Math.floor((duration - hours * 60 * 60) / 60);
		const seconds = duration - hours * 60 * 60 - minutes * 60;

		return (
			(hours > 0 ? hours + 'h ' : '') +
			(minutes > 0 || hours > 0 ? minutes + 'm ' : '') +
			(seconds + 's')
		);
	}

	renderInfo() {
		const { item } = this.props;

		const duration = item.SessionEndTime
			? this.getDurationString(
					item.SessionEndTime - item.SessionStartTime
			  )
			: 'Ongoing';

		return (
			<div className="info">
				<div className="main-info">
					<div className="name">{item.user.alias}</div>
					<div className="duration">{duration}</div>
				</div>
				<div className="start-time">
					{DateTime.fromNow(item.SessionStartTime * 1000)}
				</div>
			</div>
		);
	}

	renderMobileIcon() {
		return (
			<div className="mobile-icon">
				<div className="top" />
				<div className="screen" />
				<div className="base" />
			</div>
		);
	}

	renderAppIcon() {
		return (
			<div className="app-icon">
				<div className="screen" />
				<div className="base" />
			</div>
		);
	}

	renderValue() {
		const { item } = this.props;

		return (
			<div className="value">
				{item.appType === 'mobile'
					? this.renderMobileIcon()
					: this.renderAppIcon()}
			</div>
		);
	}

	render() {
		return (
			<div className="item">
				{this.renderImg()}
				{this.renderInfo()}
				{this.renderValue()}
			</div>
		);
	}
}

export default class RecentSessions extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			loading: true,
			pageNumber: 0,
		};
	}

	componentDidMount() {
		this.setState({ items: [] }, () => {
			this.loadData();
		});
	}

	resolveUser(userName) {
		return User.resolve({ entity: userName })
			.then(user => user)
			.catch(() => {
				return {
					alias: userName,
				};
			});
	}

	getDataWithUsers(items) {
		let requests = [];

		items.forEach(u => {
			requests.push(this.resolveUser(u.Username));
		});

		return Promise.all(requests).then(results => {
			let newItems = [];

			for (let i = 0; i < results.length; i++) {
				newItems.push({ ...items[i], user: results[i] });
			}

			return newItems;
		});
	}

	getAppType(session) {
		if (
			/android|blackberry|iphone|ipod|mobile|webos/i.test(
				session.userAgent
			)
		) {
			return 'mobile';
		}

		return 'app';
	}

	async loadData() {
		try {
			const service = await getService();
			const sessionsCollection = service.getWorkspace(ANALYTICS);
			const link = getLink(sessionsCollection, SESSIONS_LINK);
			const sessions = await service.get(link + '?limit=' + LIMIT);

			const resolvedItems = await this.getDataWithUsers(
				sessions.Items.slice(0, PAGE_SIZE)
			);

			this.setState({
				loading: false,
				allItems: sessions.Items,
				items: resolvedItems.map(x => {
					return {
						...x,
						appType: this.getAppType(x),
					};
				}),
			});
		} catch (e) {
			this.setState({ loading: false, items: [], error: e });
		}
	}

	renderHeader() {
		return (
			<div className="header">
				<div className="title">{t('title')}</div>
			</div>
		);
	}

	renderItem = (item, index) => {
		return <Item key={item.SessionStartTime} item={item} />;
	};

	renderItems() {
		const { items, loading } = this.state;

		if (loading) {
			return <Loading.Mask />;
		} else if (items && items.length === 0) {
			return <div className="no-items">{t('noItems')}</div>;
		}

		return (
			<div className="items-container">
				<div className="items">
					{(items || []).map(this.renderItem)}
				</div>
			</div>
		);
	}

	render() {
		return (
			<div className="dashboard-list-widget recent-sessions">
				{this.renderHeader()}
				{this.renderItems()}
			</div>
		);
	}
}
