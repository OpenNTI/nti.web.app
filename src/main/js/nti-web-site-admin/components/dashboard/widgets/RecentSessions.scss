@import "~nti-style-common/all.scss";

.recent-sessions {
	width: 504px;
	height: 369px;

	.title {
		margin-top: 10px;
	}

	.items-container {
		display: flex;

		.overall-stats {
			margin: 0 auto;
		}

		.items {
			width: 100%;

			.item {
				.info {
					padding-top: 0;

					.main-info {
						display: flex;

						.duration {
							margin: 2px 0 0 16px;
						}
					}
				}

				.item-image {
					border-radius: 30px;
					min-width: 40px;
				}

				.start-time, .duration {
					font-size: 12px;
					color: #b8b8b8;
				}

				.value {
					right: 16px;

					.app-icon {
						.screen {
							width: 22px;
							height: 18px;
							border-radius: 4px;
							border: solid 2px #ccc;
							text-align: center;
						}

						.base {
							width: 22px;
							height: 8px;
							background-color: #ccc;
							border-radius: 2px;
							text-align: center;
						}
					}

					.mobile-icon {
						margin-right: 2px;

						.top {
							width: 18px;
							height: 4px;
							border-radius: 4px 4px 0 0;
							background-color: #ccc;
							text-align: center;
						}

						.screen {
							width: 18px;
							height: 24px;
							border: solid 1px #ccc;
							text-align: center;
						}

						.base {
							width: 18px;
							height: 6px;
							background-color: #ccc;
							border-radius: 0 0 4px 4px;
							text-align: center;
						}
					}
				}
			}
		}
	}
}
