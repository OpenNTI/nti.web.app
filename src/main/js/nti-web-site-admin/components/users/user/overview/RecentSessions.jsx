import React from 'react';
import PropTypes from 'prop-types';
import {DateTime, Loading} from '@nti/web-commons';
import {scoped} from '@nti/lib-locale';

const LABELS = {
	title: 'Recent Sessions',
	noSessions: 'No recent sessions',
	today: 'Today'
};

const t = scoped('nti-web-site-admins.components.users.user.overview.recentsessions', LABELS);

export default class RecentSessions extends React.Component {
	static propTypes = {
		historicalSessions: PropTypes.arrayOf(PropTypes.object),
		loading: PropTypes.bool
	}

	constructor (props) {
		super(props);
	}

	renderHeader () {
		return (<div className="header"><div className="title">{t('title')}</div></div>);
	}

	renderItem = (item) => {
		const today = new Date();
		const itemDate = new Date(item.SessionStartTime * 1000);

		let dayLabel = DateTime.format(itemDate, 'dddd');

		if(DateTime.format(today, 'L') === DateTime.format(itemDate, 'L')) {
			dayLabel = t('today');
		}

		return (
			<div key={item.SessionStartTime} className="item">
				<div className="time">{DateTime.format(itemDate, 'lll')}</div>
				<div className="day">{dayLabel}</div>
			</div>
		);
	}

	renderItems () {
		const { loading, historicalSessions } = this.props;

		if(loading) {
			return <Loading.Mask/>;
		}

		if(!historicalSessions || historicalSessions.length === 0) {
			return (<div className="no-sessions">{t('noSessions')}</div>);
		}

		return (
			<div className="items">
				{historicalSessions.map(this.renderItem)}
			</div>
		);
	}

	render () {
		return (
			<div className="user-recent-sessions">
				{this.renderHeader()}
				{this.renderItems()}
			</div>
		);
	}
}
