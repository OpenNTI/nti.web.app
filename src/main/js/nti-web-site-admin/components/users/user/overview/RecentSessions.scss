@import "~@nti/style-common";

.user-recent-sessions {
	width: 241px;
	height: 340px;
	box-shadow: 0 1px 2px 1px #ccc;
	color: #555;
	font-weight: 300;
	padding: 10px;

	.header {
		.title {
			margin-left: 10px;
			font-size: 24px;
			font-weight: 500;
		}
	}

	figure.loading {
		position: relative;
		top: 20%;
	}

	.no-sessions {
		margin-left: 12px;
		margin-top: 20%;
	}

	.items {
		margin: 10px;

		.item {
			margin: 10px 0;

			.day {
				font-size: 12px;
				font-weight: 500;
				color: #b8b8b8;
			}
		}
	}
}
