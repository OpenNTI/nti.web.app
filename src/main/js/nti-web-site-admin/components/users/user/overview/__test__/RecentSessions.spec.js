import { act, create } from 'react-test-renderer';

import RecentSessions from '../RecentSessions';

/* eslint-env jest */
describe('Site admin user overview recent sessions', () => {
	test('Basic render test (2 sessions)', async () => {
		const historicalSessions = [
			{
				SessionStartTime: new Date('10/30/2017').getTime() / 1000,
			},
			{
				SessionStartTime: new Date('10/30/2017').getTime() / 1000 + 60,
			},
		];

		let cmp;
		act(() => {
			cmp = create(
				<RecentSessions historicalSessions={historicalSessions} />
			);
		});

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
		cmp.unmount();
	});

	test('Basic render test (no sessions)', async () => {
		const historicalSessions = [];

		let cmp;
		act(() => {
			cmp = create(
				<RecentSessions historicalSessions={historicalSessions} />
			);
		});

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
		cmp.unmount();
	});

	test('Basic render test (10 sessions)', async () => {
		let historicalSessions = [];

		for (let i = 0; i < 10; i++) {
			historicalSessions.push({
				SessionStartTime:
					new Date('10/30/2017').getTime() / 1000 + i * 60,
			});
		}

		let cmp;
		act(() => {
			cmp = create(
				<RecentSessions historicalSessions={historicalSessions} />
			);
		});

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
		cmp.unmount();
	});
});
