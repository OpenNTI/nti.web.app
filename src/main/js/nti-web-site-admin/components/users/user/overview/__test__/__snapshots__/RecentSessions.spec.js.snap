// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site admin user overview recent sessions Basic render test (2 sessions) 1`] = `
<div
  className="user-recent-sessions"
>
  <div
    className="header"
  >
    <div
      className="title"
    >
      Recent Sessions
    </div>
  </div>
  <div
    className="items"
  >
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:00 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:01 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
  </div>
</div>
`;

exports[`Site admin user overview recent sessions Basic render test (10 sessions) 1`] = `
<div
  className="user-recent-sessions"
>
  <div
    className="header"
  >
    <div
      className="title"
    >
      Recent Sessions
    </div>
  </div>
  <div
    className="items"
  >
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:00 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:01 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:02 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:03 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:04 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:05 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:06 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:07 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:08 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
    <div
      className="item"
    >
      <div
        className="time"
      >
        Oct. 30, 2017 12:09 AM
      </div>
      <div
        className="day"
      >
        Monday
      </div>
    </div>
  </div>
</div>
`;

exports[`Site admin user overview recent sessions Basic render test (no sessions) 1`] = `
<div
  className="user-recent-sessions"
>
  <div
    className="header"
  >
    <div
      className="title"
    >
      Recent Sessions
    </div>
  </div>
  <div
    className="no-sessions"
  >
    No recent sessions
  </div>
</div>
`;
